File: apps/api/Services/ModifierGroupService.cs
using Microsoft.EntityFrameworkCore;
using RestaurantSaas.Api.Data;
using RestaurantSaas.Api.Domain.Entities;
using RestaurantSaas.Api.Domain.Enums;
using RestaurantSaas.Api.DTOs.Menu;
using RestaurantSaas.Api.Services.Interfaces;

namespace RestaurantSaas.Api.Services;

public class ModifierGroupService(AppDbContext db) : IModifierGroupService
{
    // ─── Queries ───────────────────────────────────────────────────────────────

    public async Task<IEnumerable<ModifierGroupDto>> GetGroupsAsync(
        Guid restaurantId, Guid? branchId = null)
    {
        var query = db.ModifierGroups
            .Include(g => g.Branch)
            .Include(g => g.Options)
            .Where(g => g.RestaurantId == restaurantId && g.IsActive);

        if (branchId.HasValue)
            query = query.Where(g => g.BranchId == branchId.Value);

        var groups = await query
            .OrderBy(g => g.SortOrder)
            .ToListAsync();

        return groups.Select(g => ToDto(g));
    }

    public async Task<ModifierGroupDto?> GetGroupAsync(
        Guid id, Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .Include(g => g.Branch)
            .Include(g => g.Options)
            .FirstOrDefaultAsync(g =>
                g.Id == id &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        return group is null ? null : ToDto(group);
    }

    // ─── Mutations ─────────────────────────────────────────────────────────────

    public async Task<(ModifierGroupDto? Group, string? Error)> CreateGroupAsync(
        CreateModifierGroupRequest request, Guid restaurantId, Guid branchId)
    {
        var branch = await db.Branches
            .FirstOrDefaultAsync(b => b.Id == branchId && b.RestaurantId == restaurantId);
        if (branch is null) return (null, "BRANCH_NOT_FOUND");

        if (!Enum.TryParse<SelectionType>(request.SelectionType, out var selectionType))
            return (null, "INVALID_SELECTION_TYPE");

        var group = new ModifierGroup
        {
            Id            = Guid.NewGuid(),
            RestaurantId  = restaurantId,
            BranchId      = branchId,
            Name          = request.Name.Trim(),
            SelectionType = selectionType,
            IsRequired    = request.IsRequired,
            MinSelections = request.MinSelections,
            MaxSelections = request.MaxSelections,
            SortOrder     = request.SortOrder,
            IsActive      = true,
            CreatedAt     = DateTime.UtcNow,
            Branch        = branch,
            Options       = []
        };

        db.ModifierGroups.Add(group);
        await db.SaveChangesAsync();

        return (ToDto(group), null);
    }

    public async Task<(ModifierGroupDto? Group, string? Error)> UpdateGroupAsync(
        Guid id, UpdateModifierGroupRequest request, Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .Include(g => g.Branch)
            .Include(g => g.Options)
            .FirstOrDefaultAsync(g =>
                g.Id == id &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        if (group is null) return (null, "NOT_FOUND");

        if (!Enum.TryParse<SelectionType>(request.SelectionType, out var selectionType))
            return (null, "INVALID_SELECTION_TYPE");

        group.Name          = request.Name.Trim();
        group.SelectionType = selectionType;
        group.IsRequired    = request.IsRequired;
        group.MinSelections = request.MinSelections;
        group.MaxSelections = request.MaxSelections;
        group.SortOrder     = request.SortOrder;
        group.IsActive      = request.IsActive;

        await db.SaveChangesAsync();
        return (ToDto(group), null);
    }

    public async Task<(bool Success, string? Error)> DeactivateGroupAsync(
        Guid id, Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .FirstOrDefaultAsync(g =>
                g.Id == id &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        if (group is null) return (false, "NOT_FOUND");

        var hasOptions = await db.ModifierOptions
            .AnyAsync(o => o.ModifierGroupId == id && o.IsActive);
        if (hasOptions) return (false, "GROUP_HAS_OPTIONS");

        var hasLinks = await db.ProductVariantModifierGroups
            .AnyAsync(pvmg => pvmg.ModifierGroupId == id);
        if (hasLinks) return (false, "GROUP_HAS_LINKS");

        group.IsActive = false;
        await db.SaveChangesAsync();
        return (true, null);
    }

    // ─── Options ───────────────────────────────────────────────────────────────

    public async Task<(ModifierOptionDto? Option, string? Error)> AddOptionAsync(
        Guid groupId, CreateModifierOptionRequest request, Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .FirstOrDefaultAsync(g =>
                g.Id == groupId &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        if (group is null) return (null, "GROUP_NOT_FOUND");

        var option = new ModifierOption
        {
            Id              = Guid.NewGuid(),
            ModifierGroupId = groupId,
            Name            = request.Name.Trim(),
            PriceDelta      = request.PriceDelta,
            IsDefault       = request.IsDefault,
            IsActive        = true,
            SortOrder       = request.SortOrder
        };

        db.ModifierOptions.Add(option);
        await db.SaveChangesAsync();

        return (ToOptionDto(option), null);
    }

    public async Task<(ModifierOptionDto? Option, string? Error)> UpdateOptionAsync(
        Guid groupId, Guid optionId, UpdateModifierOptionRequest request,
        Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .FirstOrDefaultAsync(g =>
                g.Id == groupId &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        if (group is null) return (null, "GROUP_NOT_FOUND");

        var option = await db.ModifierOptions
            .FirstOrDefaultAsync(o => o.Id == optionId && o.ModifierGroupId == groupId);

        if (option is null) return (null, "OPTION_NOT_FOUND");

        option.Name       = request.Name.Trim();
        option.PriceDelta = request.PriceDelta;
        option.IsDefault  = request.IsDefault;
        option.SortOrder  = request.SortOrder;
        option.IsActive   = request.IsActive;

        await db.SaveChangesAsync();
        return (ToOptionDto(option), null);
    }

    public async Task<(bool Success, string? Error)> DeactivateOptionAsync(
        Guid groupId, Guid optionId, Guid restaurantId, Guid? branchId = null)
    {
        var group = await db.ModifierGroups
            .FirstOrDefaultAsync(g =>
                g.Id == groupId &&
                g.RestaurantId == restaurantId &&
                (branchId == null || g.BranchId == branchId.Value));

        if (group is null) return (false, "GROUP_NOT_FOUND");

        var option = await db.ModifierOptions
            .FirstOrDefaultAsync(o => o.Id == optionId && o.ModifierGroupId == groupId);

        if (option is null) return (false, "OPTION_NOT_FOUND");

        option.IsActive = false;
        await db.SaveChangesAsync();
        return (true, null);
    }

    // ─── Helpers ───────────────────────────────────────────────────────────────

    private static ModifierGroupDto ToDto(ModifierGroup g) => new(
        g.Id,
        g.BranchId,
        g.Branch?.Name ?? string.Empty,
        g.Name,
        g.SelectionType.ToString(),
        g.IsRequired,
        g.MinSelections,
        g.MaxSelections,
        g.SortOrder,
        g.IsActive,
        g.CreatedAt,
        g.Options.Where(o => o.IsActive).OrderBy(o => o.SortOrder)
            .Select(o => new ModifierOptionDto(o.Id, o.Name, o.PriceDelta, o.IsDefault, o.IsActive, o.SortOrder))
    );

    private static ModifierOptionDto ToOptionDto(ModifierOption o) =>
        new(o.Id, o.Name, o.PriceDelta, o.IsDefault, o.IsActive, o.SortOrder);
}

File: apps/web/src/app/core/services/modifier-group.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiClientService } from './api-client.service';

export interface ModifierOptionDto {
  id: string;
  name: string;
  priceDelta: number;
  isDefault: boolean;
  isActive: boolean;
  sortOrder: number;
}

export interface ModifierGroupDto {
  id: string;
  branchId: string;
  branchName: string;
  name: string;
  selectionType: 'Single' | 'Multiple';
  isRequired: boolean;
  minSelections: number;
  maxSelections: number | null;
  sortOrder: number;
  isActive: boolean;
  createdAt: string;
  options: ModifierOptionDto[];
}

export interface CreateModifierGroupRequest {
  branchId: string | null;
  name: string;
  selectionType: string;
  isRequired: boolean;
  minSelections: number;
  maxSelections: number | null;
  sortOrder: number;
}

export interface UpdateModifierGroupRequest {
  name: string;
  selectionType: string;
  isRequired: boolean;
  minSelections: number;
  maxSelections: number | null;
  sortOrder: number;
  isActive: boolean;
}

export interface CreateModifierOptionRequest {
  name: string;
  priceDelta: number;
  isDefault: boolean;
  sortOrder: number;
}

export interface UpdateModifierOptionRequest {
  name: string;
  priceDelta: number;
  isDefault: boolean;
  sortOrder: number;
  isActive: boolean;
}

@Injectable({ providedIn: 'root' })
export class ModifierGroupService {
  constructor(private api: ApiClientService) {}

  getGroups(): Observable<ModifierGroupDto[]> {
    return this.api.get<ModifierGroupDto[]>('/modifier-groups');
  }

  getGroup(id: string): Observable<ModifierGroupDto> {
    return this.api.get<ModifierGroupDto>(`/modifier-groups/${id}`);
  }

  createGroup(request: CreateModifierGroupRequest): Observable<ModifierGroupDto> {
    return this.api.post<ModifierGroupDto>('/modifier-groups', request);
  }

  updateGroup(id: string, request: UpdateModifierGroupRequest): Observable<ModifierGroupDto> {
    return this.api.put<ModifierGroupDto>(`/modifier-groups/${id}`, request);
  }

  deactivateGroup(id: string): Observable<void> {
    return this.api.delete<void>(`/modifier-groups/${id}`);
  }

  addOption(groupId: string, request: CreateModifierOptionRequest): Observable<ModifierOptionDto> {
    return this.api.post<ModifierOptionDto>(`/modifier-groups/${groupId}/options`, request);
  }

  updateOption(groupId: string, optionId: string, request: UpdateModifierOptionRequest): Observable<ModifierOptionDto> {
    return this.api.put<ModifierOptionDto>(`/modifier-groups/${groupId}/options/${optionId}`, request);
  }

  deactivateOption(groupId: string, optionId: string): Observable<void> {
    return this.api.delete<void>(`/modifier-groups/${groupId}/options/${optionId}`);
  }
}
